macros/Compat.ejs

```diff
@@ -1,66 +1,399 @@
+var output = '';
+
+var compatStrings = string.deserialize(template("L10n:CompatTable"));
+var localize = mdn.getLocalString;
+
+var s_no_data_found = mdn.localString({
+    'en-US': `No compatibility data found. Please contribute data for "${query}" to the <a href="https://github.com/mdn/browser-compat-data">MDN compatibility data repository</a>.`
 });
 
+const desktopBrowsers = {
+  chrome: "Chrome",
+  firefox: "Firefox",
+  edge: "Edge",
+  ie: "Internet Explorer",
+  opera: "Opera",
+  safari: "Safari",
+  servo: "Servo"
+};
+
+const mobileBrowsers = {
+  webview_android: "Android",
+  chrome_android: "Chrome for Android",
+  edge_mobile: "Edge mobile",
+  firefox_android: "Firefox for Android",
+  ie_mobile: "IE mobile",
+  opera_android: "Opera Android",
+  safari_ios: "iOS Safari",
+}
+
+var notesArray = [];
 
-if ($0) {
-  try {
-    var data = mdn.fetchJSONResource(url);
-  } catch (e) {
-    output = 'File "' + file + '" not found or malformed.'
+/*
+Write the table header.
+*/
+function writeTableHead(environment, browserNames) {
+  let browserNameKeys = Object.keys(browserNames);
+  let output = `<div id="compat-${environment}"><table class="compat-table"><thead><tr>`;
+  output +=  `<th>${localize(compatStrings, 'feature')}</th>`
+  for (let browserNameKey of browserNameKeys) {
+    output += `<th>${browserNames[browserNameKey]}</th>`;
   }
+  output += "</tr></thead>";
+  return output;
 }
 
-if (data) {
-    if (leaf !== '') {
-      output = template("CompatDataTable", [data[leaf], "api"]);
+/*
+Given the value of `version_added` or `version_removed`, this returns:
+  * a string to appear in the table cell, like "Yes", "No" or "?"
+
+*/
+function getVersionString(versionInfo) {
+  switch (versionInfo) {
+    case null:
+      return `<span title="${localize(compatStrings, 'supportsShort_unknown_title')}"
+              style="color: rgb(255, 153, 0);">${localize(compatStrings, 'supportsShort_unknown')}</span>`;
+      break;
+    case true:
+      return `<span style="color: #888">(${localize(compatStrings, 'supportsShort_yes')})</span>`;
+      break;
+    case false:
+      return `<span style="color: #f00">(${localize(compatStrings, 'supportsShort_no')})</span>`;
+      break;
+    default:
+      return versionInfo;
+  }
+}
+
+/*
+  Create the note for a browser flag or preference
+*/
+function generateFlagsNote(supportData, browserId) {
+  var versions = '';
+  if (typeof(supportData.version_added) === 'string') {
+    versions = 'From version ' + supportData.version_added;
+  }
+  if (typeof(supportData.version_removed) === 'string') {
+    if (versions) {
+      versions += ' until version '+ supportData.version_removed + ' (excluded), ';
     } else {
-      output = template("CompatDataTable", [data, "module"]);
+      versions = 'Until version ' + supportData.version_removed + ' (excluded), ';
     }
+  } else {
+    versions += ', ';
+  } if ((browserId === 'firefox') || (browserId === 'firefox_android')) {
```

Comment:
No longer relevant in the refactored function. I changed this to a switch statement, as I expect that there are more browsers with preferences that we want to add once we stumble upon a case.